Tools/LambdaTestTool-v2/src/Amazon.Lambda.TestTool/Extensions/HttpContextExtensions.cs

```diff
@@ -0,0 +1,214 @@
+        {
+            // this should be decoded value
+            httpApiV2ProxyRequest.PathParameters = pathParameters;
+        }
+
+        if (HttpRequestUtility.IsBinaryContent(request.ContentType))
+        {
+            // we already converted it when we read the body so we dont need to re-convert it
+            httpApiV2ProxyRequest.IsBase64Encoded = true;
+        }
+
+        return httpApiV2ProxyRequest;
+    }
+
+    /// <summary>
+    /// Translates an <see cref="HttpContext"/> to an <see cref="APIGatewayProxyRequest"/>.
+    /// </summary>
+    /// <param name="context">The <see cref="HttpContext"/> to be translated.</param>
+    /// <param name="apiGatewayRouteConfig">The configuration of the API Gateway route, including the HTTP method, path, and other metadata.</param>
+    /// <returns>An <see cref="APIGatewayProxyRequest"/> object representing the translated request.</returns>
+    public static async Task<APIGatewayProxyRequest> ToApiGatewayRequest(
+        this HttpContext context,
+        ApiGatewayRouteConfig apiGatewayRouteConfig,
+        ApiGatewayEmulatorMode emulatorMode)
+    {
+        var request = context.Request;
+        var body = await HttpRequestUtility.ReadRequestBody(request);
+        var contentLength = HttpRequestUtility.CalculateContentLength(request, body);
+
+        var pathParameters = RouteTemplateUtility.ExtractPathParameters(apiGatewayRouteConfig.Path, request.Path);
+
+        var (headers, multiValueHeaders) = HttpRequestUtility.ExtractHeaders(request.Headers);
+        var (queryStringParameters, multiValueQueryStringParameters) = HttpRequestUtility.ExtractQueryStringParameters(request.Query);
+
+        if (!headers.ContainsKey("content-length") && emulatorMode != ApiGatewayEmulatorMode.Rest) // rest doesnt set content-length by default
+        {
+            headers["content-length"] = contentLength.ToString();
+            multiValueHeaders["content-length"] = [contentLength.ToString()];
+        }
+
+        if (!headers.ContainsKey("content-type"))
+        {
+            headers["content-type"] = "text/plain; charset=utf-8";
+            multiValueHeaders["content-type"] = ["text/plain; charset=utf-8"];
+        }
+
+        // This is the decoded value
+        var path = request.Path.Value;
+
+        if (emulatorMode == ApiGatewayEmulatorMode.HttpV1 || emulatorMode == ApiGatewayEmulatorMode.Rest) // rest and httpv1 uses the encoded value for path an
+        {
+            path = request.Path.ToUriComponent();
+        }
+
+        if (emulatorMode == ApiGatewayEmulatorMode.Rest) // rest uses encoded value for the path params
+        {
+            var encodedPathParameters = pathParameters.ToDictionary(
+                    kvp => kvp.Key,
+                    kvp => Uri.EscapeUriString(kvp.Value)); // intentionally using EscapeURiString over EscapeDataString since EscapeURiString correctly handles reserved characters :/?#[]@!$&'()*+,;= in this case
```

Comment:
for example. escapedatastring encodes * as %20 which is different than how api gateway rest api does it (it keeps * as *)